README.md
# Entity Matching model

[![emm package in P11945-outgoing feed in Azure
Artifacts](https://feeds.dev.azure.com/INGCDaaS/49255723-5232-4e9f-9501-068bf5e381a9/_apis/public/Packaging/Feeds/P11945-outgoing/Packages/8436e3e5-0029-4c5e-9a98-a9961acdd9a0/Badge)](https://dev.azure.com/INGCDaaS/IngOne/_artifacts/feed/P11945-outgoing/PyPI/emm?preferRelease=true)

Entity Matching Model (EMM) solves the problem of matching company names between two possibly very
large datasets. EMM can match millions against millions of names with a distributed approach.
It uses the well-established candidate selection techniques in string matching,
namely: tfidf vectorization combined with cosine similarity (with significant optimization),
both word-based and character-based, and sorted neighbourhood indexing.
These so-called indexers act complementary for selecting realistic name-pair candidates.
On top of the indexers, EMM has a classifier with optimized string-based, rank-based, and legal-entity
based features to estimate how confident a company name match is.

The classifier can be trained to give a string similarity score or a probability of match.
Both types of score are useful, in particular when there are many good-looking matches to choose between.
Optionally, the EMM package can also be used to match a group of company names that belong together,
to a common company name in the ground truth. For example, all different names used to address an external bank account.
This step aggregates the name-matching scores from the supervised layer into a single match.

The package is modular in design and and works both using both Pandas and Spark. A classifier trained with the former
can be used with the latter and vice versa.

For release history see: ``CHANGES.md``.

## Notebooks

For detailed examples of the code please see the notebooks under `notebooks/`.

- `01-entity-matching-pandas-version.ipynb`: Using the Pandas version of EMM for name-matching.
- `02-entity-matching-spark-version.ipynb`: Using the Spark version of EMM for name-matching.
- `03-entity-matching-training-pandas-version.ipynb`: Fitting the supervised model and setting a discrimination threshold (Pandas).
- `04-entity-matching-aggregation-pandas-version.ipynb`: Using the aggregation layer and setting a discrimination threshold (Pandas).

## Documentation

For documentation, design, and API see `docs/`.


## Check it out

The Entity matching model library requires Python >= 3.7 and is pip friendly. To get started, simply do:

```shell
pip install emm
```

or check out the code from our repository:

```shell
git clone https://github.com/ing-bank/EntityMatchingModel.git
pip install -e EntityMatchingModel/
```

where in this example the code is installed in edit mode (option -e).

Additional dependencies can be installed with, e.g.:

```shell
pip install "emm[spark,dev,test]"
```

You can now use the package in Python with:


```python
import emm
```

**Congratulations, you are now ready to use the Entity Matching model!**

## Quick run

As a quick example, you can do:

```python
from emm import PandasEntityMatching
from emm.data.create_data import create_example_noised_names

# generate example ground-truth names and matching noised names, with typos and missing words.
ground_truth, noised_names = create_example_noised_names(random_seed=42)
train_names, test_names = noised_names[:5000], noised_names[5000:]

# two example name-pair candidate generators: character-based cosine similarity and sorted neighbouring indexing
indexers = [
  {
      'type': 'cosine_similarity',
      'tokenizer': 'characters',   # character-based cosine similarity. alternative: 'words'
      'ngram': 2,                  # 2-character tokens only
      'num_candidates': 5,         # max 5 candidates per name-to-match
      'cos_sim_lower_bound': 0.2,  # lower bound on cosine similarity
  },
  {'type': 'sni', 'window_length': 3}  # sorted neighbouring indexing window of size 3.
]
em_params = {
  'name_only': True,         # only consider name information for matching
  'entity_id_col': 'Index',  # important to set both index and name columns to pick up
  'name_col': 'Name',
  'indexers': indexers,
  'supervised_on': False,    # no supervided model (yet) to select best candidates
  'with_legal_entity_forms_match': True,   # add feature that indicates match of legal entity forms (e.g. ltd != co)
}
# 1. initialize the entity matcher
p = PandasEntityMatching(em_params)

# 2. fitting: prepare the indexers based on the ground truth names, eg. fit the tfidf matrix of the first indexer.
p.fit(ground_truth)

# 3. create and fit a supervised model for the PandasEntityMatching object, to pick the best match (this takes a while)
#    input is "positive" names column 'Name' that are all supposed to match to the ground truth,
#    and an id column 'Index' to check with candidate name-pairs are matching and which not.
#    A fraction of these names may be turned into negative names (= no match to the ground truth).
#    (internally, candidate name-pairs are automatically generated, these are the input to the classification)
p.fit_classifier(train_names, create_negative_sample_fraction=0.5)

# 4. scoring: generate pandas dataframe of all name-pair candidates.
#    The classifier-based probability of match is provided in the column 'nm_score'.
#    Note: can also call p.transform() without training the classifier first.
candidates_scored_pd = p.transform(test_names)

# 5. scoring: for each name-to-match, select the best ground-truth candidate.
best_candidates = candidates_scored_pd[candidates_scored_pd.best_match]
best_candidates.head()
```

For Spark, you can use the class `SparkEntityMatching` instead, with the same API as the Pandas version.
For all available examples, please see the tutorial notebooks under `notebooks/`.

## Project contributors

This package was authored by ING Analytics Wholesale Banking.

## Contact and support

Contact the WBAA team via Github issues.
Please note that INGA-WB provides support only on a best-effort basis.

## License

Copyright ING WBAA 2023. Entity Matching Model is completely free, open-source and licensed under the [MIT license](https://en.wikipedia.org/wiki/MIT_License).


CHANGES.md
# Release notes

## Version 1.4.1

- Major updates of documentation for open-sourcing
- Add extra_features option to emm.fit_classifier function
- Add option drop_duplicate_candidates option to prepare_name_pairs_pd function
- Rename SupervisedLayerEstimator as SparkSupervisedLayerEstimator
- Consistent carry_on_cols behavior between pandas and spark indexing classes
- Significant cleanup of cleanup of parameters.py.
- Remove init_spark file and related calls
- Cleanup of util and spark_utils functions
- Remove unused dap related io functions

## Version 1.4.0

- Introduce `Timer` context for logging
- Removed backwards compatibility `unionByName` helper. Spark >= 3.1 required.
- Replaced custom "NON NFKD MAP" with `unidecode`
- Integration test speedup: split-off long-running integration test
- Removed: `verbose`, `compute_missing`, `use_tqdm`, `save_intermediary`, `n_jobs` options removed, `mlflow` dependencies
- Removed: prediction explanations (bloat), unused unsupervised model, "name_clustering" aggregation
- Perf: 5-10x speedup of feature computations
- Perf: `max_frequency_nm_score` and `mean_score` aggregation method short-circuit groups with only one record (2-3x speedup for skewed datasets)
- Tests: added requests retries with backoff for unstable connections

## Version 1.3.14

- Converted RST readme and changelog to Markdown
- Introduced new parameters for force execution and cosine similary threads.

## Version 1.3.5-1.3.13

See git history for changes.

## Version 1.3.4, Jan 2023

- Added helper function to activate mlflow tracking.
- Added spark example to example.py
- Minor updates to documentation.

## Version 1.3.3, Dec 2022

- Added sm feature indicating matches of legal entity forms between names. Turn on with parameter
 `with_legal_entity_forms_match=True`. Example usage in:
    `03-entity-matching-training-pandas-version.ipynb`. For
    code see `calc_features/cleanco_lef_matching.py`.
- Added code for calculating discrimination threshold curves:
    `em.calc_threshold()`. Example usage in:
    `03-entity-matching-training-pandas-version.ipynb`.
- Added example notebook for name aggregation. See:
    `04-entity-matching-aggregation-pandas-version.ipynb`.


example.py
# Copyright (c) 2023 ING Analytics Wholesale Banking
#
# Permission is hereby granted, free of charge, to any person obtaining a copy of
# this software and associated documentation files (the "Software"), to deal in
# the Software without restriction, including without limitation the rights to
# use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
# the Software, and to permit persons to whom the Software is furnished to do so,
# subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
# FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
# COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
# IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

import pandas as pd

from emm import PandasEntityMatching
from emm.data.create_data import create_example_noised_names
from emm.helper import spark_installed

if spark_installed:
    from emm import SparkEntityMatching


def example():
    """Simple entity matching example using PandasEntityMatching"""
    # This is the example shown in the readme.
    # if you update this example, please update the readme and vice versa!

    # generate example ground truth names and matching noised names, with typos and missing words
    ground_truth, noised_names = create_example_noised_names(random_seed=43)
    train_names, test_names = noised_names[:5000], noised_names[5000:]

    # two example name-pair candidate generators: character-based cosine similarity and sorted neighbouring indexing
    indexers = [
        {
            "type": "cosine_similarity",
            "tokenizer": "characters",  # character-based cosine similarity
            "ngram": 2,  # 2-gram tokens only
            "num_candidates": 5,  # max 5 candidates per name-to-match
            "cos_sim_lower_bound": 0.2,  # lower bound on cosine similarity
        },
        {
            "type": "sni",
            "window_length": 3,
        },  # sorted neighbouring indexing window of size 3.
    ]
    em_params = {
        "name_only": True,  # only consider name information for matching
        "entity_id_col": "Index",  # important to set both index and name columns
        "name_col": "Name",
        "indexers": indexers,
        "supervised_on": False,  # no initial supervised model to select best candidates right now
        "with_legal_entity_forms_match": True,  # add feature that indicates match of legal entity forms (eg. ltd != co)
    }
    # initialize the entity matcher
    p = PandasEntityMatching(em_params)
    # prepare the indexers based on the ground truth names: e.g. fit the tfidf matrix of the first indexer.
    p.fit(ground_truth)

    # pandas dataframe with name-pair candidates, made by the indexers. all names have been preprocessed.
    candidates_pd = p.transform(test_names)
    candidates_pd.head()

    # create and fit a supervised model for the PandasEntityMatching object to pick the best match (this takes a while)
    # input is "positive" names column 'Name' that are all supposed to match to the ground truth,
    # and an id column 'Index' to check with candidate name-pairs are matching and which not.
    # A fraction of these names may be turned into negative names (no match to the ground truth).
    # (internally candidate name-pairs are automatically generated, which are input for the classification)
    p.fit_classifier(train_positive_names_to_match=train_names, create_negative_sample_fraction=0.5)

    # generated name-pair candidates, now with classifier-based probability of match.
    # Input is the names' column 'Name'. In the output candidates df, see extra column 'nm_score'.
    candidates_scored_pd = p.transform(test_names)
    candidates_scored_pd.head()

    # for each name-to-match, select the best ground-truth candidate
    best_candidates = candidates_scored_pd[candidates_scored_pd.best_match].copy()

    # print some performance statistics (which is possible in this example as we know the correct match).
    best_candidates["correct"] = best_candidates["gt_entity_id"] == best_candidates["entity_id"]
    print(f"Number of names-to-match: {len(test_names)}")
    print(f"Number of best candidates: {len(best_candidates)}")
    print(f"Number of correct matches: {len(best_candidates[best_candidates.correct])}")
    print(f"Number of incorrect matches: {len(best_candidates[~best_candidates.correct])}")

    # return these numbers for unit-testing
    n_ground_truth = len(ground_truth)
    n_noised_names = len(noised_names)
    n_names_to_match = len(test_names)
    n_best_match = len(best_candidates)
    n_correct = len(best_candidates[best_candidates.correct])
    n_incorrect = len(best_candidates[~best_candidates.correct])

    return (
        n_ground_truth,
        n_noised_names,
        n_names_to_match,
        n_best_match,
        n_correct,
        n_incorrect,
    )


def example_pandas():
    """Simple pandas entity matching example using PandasEntityMatching"""
    # Another example, but this time in pandas with dummy ground truth and names-to-match.
    # (Otherwise same settings as the pandas example above.)

    ground_truth = pd.DataFrame(
        {"name": ["Apple", "Microsoft", "Google", "Amazon", "Netflix", "Spotify"], "id": [1, 2, 3, 4, 5, 6]}
    )
    train_names = pd.DataFrame(
        {"name": ["MicorSoft", "Gugle", "Netfliks", "Spot-on", "Spot-off"], "id": [2, 3, 5, 6, 6]}
    )
    test_names = pd.DataFrame(
        {"name": ["Apl", "Aplle", "Microbloft", "Netflfli", "amz", "googol"], "id": [1, 1, 2, 5, 4, 3]}
    )

    # two example name-pair candidate generators: character-based cosine similarity and sorted neighbouring indexing
    indexers = [
        {
            "type": "cosine_similarity",
            "tokenizer": "characters",  # character-based cosine similarity
            "ngram": 2,  # 2-gram tokens only
            "num_candidates": 5,  # max 5 candidates per name-to-match
            "cos_sim_lower_bound": 0.2,  # lower bound on cosine similarity
        },
        {
            "type": "sni",
            "window_length": 3,
        },  # sorted neighbouring indexing window of size 3.
    ]
    emm_config = {
        "name_only": True,  # only consider name information for matching
        "entity_id_col": "id",  # important to set both index and name columns
        "name_col": "name",
        "indexers": indexers,
        "supervised_on": False,  # no initial supervised model to select best candidates right now
    }

    # fitting of first the ground truth, then the training names to match.
    model = PandasEntityMatching(emm_config)
    model.fit(ground_truth)
    model.fit_classifier(train_names, create_negative_sample_fraction=0.5)

    candidates_scored = model.transform(test_names)

    best_candidates = candidates_scored[candidates_scored.score_0 > 0][["name", "gt_name", "gt_entity_id"]]

    best_candidates.head()
    """
    +----------+---------+------------+
    |      name|  gt_name|gt_entity_id|
    +----------+---------+------------+
    |       Apl|    Apple|           1|
    |     Aplle|    Apple|           1|
    |Microbloft|Microsoft|           2|
    |  Netflfli|  Netflix|           5|
    |       amz|   Amazon|           4|
    |    googol|   Google|           3|
    +----------+---------+------------+
    """
    # return dataframe for unit-testing
    return best_candidates


def example_spark(spark):
    """Simple spark entity matching example using SparkEntityMatching"""
    # Another example, but this time in spark, with dummy ground truth and names-to-match.
    # (Otherwise same settings as the pandas example above.)

    ground_truth = spark.createDataFrame(
        [
            ("Apple", 1),
            ("Microsoft", 2),
            ("Google", 3),
            ("Amazon", 4),
            ("Netflix", 5),
            ("Spotify", 6),
        ],
        ["name", "id"],
    )
    train_names = spark.createDataFrame(
        [
            ("MicorSoft", 2),
            ("Gugle", 3),
            ("Netfliks", 5),
            ("Spot-on", 6),
            ("Spot-off", 6),
        ],
        ["name", "id"],
    )
    test_names = spark.createDataFrame(
        [
            ("Apl", 1),
            ("Aplle", 1),
            ("Microbloft", 2),
            ("Netflfli", 5),
            ("amz", 4),
            ("googol", 3),
        ],
        ["name", "id"],
    )

    # two example name-pair candidate generators: character-based cosine similarity and sorted neighbouring indexing
    indexers = [
        {
            "type": "cosine_similarity",
            "tokenizer": "characters",  # character-based cosine similarity
            "ngram": 2,  # 2-gram tokens only
            "num_candidates": 5,  # max 5 candidates per name-to-match
            "cos_sim_lower_bound": 0.2,  # lower bound on cosine similarity
        },
        {
            "type": "sni",
            "window_length": 3,
        },  # sorted neighbouring indexing window of size 3.
    ]
    emm_config = {
        "name_only": True,  # only consider name information for matching
        "entity_id_col": "id",  # important to set both index and name columns
        "name_col": "name",
        "indexers": indexers,
        "supervised_on": False,  # no initial supervised model to select best candidates right now
    }

    # fitting of first the ground truth, then the training names to match.
    model = SparkEntityMatching(emm_config)
    model.fit(ground_truth)
    model.fit_classifier(train_names, create_negative_sample_fraction=0.5)

    candidates_scored = model.transform(test_names)

    best_candidates = candidates_scored.where(candidates_scored.score_0 > 0).select("name", "gt_name", "gt_entity_id")

    best_candidates.show()
    """
    +----------+---------+------------+
    |      name|  gt_name|gt_entity_id|
    +----------+---------+------------+
    |       Apl|    Apple|           1|
    |     Aplle|    Apple|           1|
    |Microbloft|Microsoft|           2|
    |  Netflfli|  Netflix|           5|
    |       amz|   Amazon|           4|
    |    googol|   Google|           3|
    +----------+---------+------------+
    """
    # return dataframe for unit-testing
    return best_candidates.toPandas()
